dbms/src/Storages/DeltaMerge/VersionChain/NewHandleIndex.h

```diff
@@ -0,0 +1,142 @@
+// In order to save memory, it stores the hash value of the handle instead of the handle itself.
+template <ExtraHandleType HandleType, typename Hash = absl::Hash<typename ExtraHandleRefType<HandleType>::type>>
+class NewHandleIndex
+{
+public:
+    using HandleRefType = typename ExtraHandleRefType<HandleType>::type;
+
+    // Different handles may have hash collisions, so during the lookup process, after finding a matching hash value,
+    // it is necessary to use the delta_reader to read the actual data for verification.
+    std::optional<RowID> find(HandleRefType handle, DeltaValueReader & delta_reader, const UInt32 stable_rows) const
+    {
+        const auto hash_value = hasher(handle);
+        const auto [start, end] = handle_to_row_id.equal_range(hash_value);
+        if (start == end)
+            return {};
+
+        MutableColumns mut_cols(1);
+        mut_cols[0] = createHandleColumn();
+        mut_cols[0]->reserve(std::distance(start, end));
+        size_t total_read_rows = 0;
+        for (auto itr = start; itr != end; ++itr)
+        {
+            // TODO: Maybe we can support readRows in batch.
+            // But hash collision is rare.
+            // And the distribution of these conflicting values is typically scattered.
+            // The benefits should be quite small, so it's not a high priority.
+            const auto read_rows = delta_reader.readRows(
+                mut_cols,
+                /*offset*/ itr->second - stable_rows,
+                /*limit*/ 1,
+                /*range*/ nullptr);
+            RUNTIME_CHECK(read_rows == 1, itr->second, stable_rows, read_rows);
+            ++total_read_rows;
+            ColumnView<HandleType> handles(*(mut_cols[0]));
+            if (handles[total_read_rows - 1] == handle)
+                return itr->second;
+        }
+        return {};
+    }
+
+    void insert(HandleRefType handle, RowID row_id) { std::ignore = handle_to_row_id.emplace(hasher(handle), row_id); }
+
+    // The hash value of a handle cannot reflect the order of handles,
+    // so data can only be deleted by traversing the entire structure.
+    // First, collect all row_ids in the index and sort them.
+    // Then, read the handles in batches from the delta, and check if the handles in the delete range.
+    void deleteRange(const RowKeyRange & range, DeltaValueReader & delta_reader, const UInt32 stable_rows)
+    {
+        if (handle_to_row_id.empty())
+            return;
+
+        std::vector<RowID> row_ids;
+        row_ids.reserve(handle_to_row_id.size());
+        for (auto [h, row_id] : handle_to_row_id)
+            row_ids.push_back(row_id - stable_rows);
+
+        std::sort(row_ids.begin(), row_ids.end());
+
+        absl::btree_multimap<UInt32, RowID> t;
+        auto begin = row_ids.begin();
+        auto end = row_ids.end();
+        auto get_next_continuous_size = [](auto begin, auto end) {
```

Comment:
How long is the `continuous size` expected to be? And how often the `RowId`s are not continuous?

If the `RowId`s are mostly continuous, then is there any chance we can iterate from the tail?